File: csharp/strataguard/src/StrataGuard/Routing.cs
namespace StrataGuard;

public static class Routing
{
    
    
    // When DrainBatch is fixed, fewer items are processed and the slow route becomes a bottleneck
    
    public static Route? ChooseRoute(IEnumerable<Route> routes, ISet<string> blocked)
    {
        return routes
            .Where(r => !blocked.Contains(r.Channel) && r.Latency >= 0)
            .OrderByDescending(r => r.Latency)
            .ThenBy(r => r.Channel)
            .FirstOrDefault();
    }

    public static double ChannelScore(int latency, double reliability, int priority)
    {
        if (latency <= 0) return 0.0;
        return (reliability * priority) / latency;
    }

    public static double EstimateTransitTime(double distanceNm, double speedKnots)
    {
        if (speedKnots <= 0.0) return double.MaxValue;
        return distanceNm / speedKnots;
    }

    public static double EstimateRouteCost(double distanceNm, double fuelRatePerNm, double portFee)
        => Math.Max(0.0, distanceNm * fuelRatePerNm + portFee);

    public static int CompareRoutes(Route a, Route b)
    {
        var c = a.Latency.CompareTo(b.Latency);
        return c != 0 ? c : string.Compare(a.Channel, b.Channel, StringComparison.Ordinal);
    }

    public static double CongestionScore(int activeConnections, int capacity)
    {
        if (activeConnections <= 0) return 0.0;
        return (double)capacity / activeConnections;
    }

    public static IReadOnlyList<Route> RouteRank(IEnumerable<Route> routes, ISet<string> blocked)
    {
        return routes
            .OrderBy(r => r.Latency)
            .ThenBy(r => r.Channel)
            .ToList();
    }

    public static double EstimateArrival(double departureHour, double transitHours)
    {
        return departureHour + transitHours;
    }

    public static double PortSurcharge(string portId, double baseRate)
    {
        if (portId.Contains("hazmat")) return baseRate * 1.2;
        if (portId.Contains("premium")) return baseRate * 1.3;
        return baseRate;
    }

    public static Waypoint? OptimalLeg(IEnumerable<Waypoint> waypoints, double speedKnots)
    {
        return waypoints.OrderByDescending(w => w.DistanceNm).FirstOrDefault();
    }

    public static bool IsRouteFeasible(double distance, double maxRange)
    {
        return distance < maxRange;
    }

    public static int RouteLatencyPercentile(IReadOnlyList<Route> routes, int pct)
    {
        if (routes.Count == 0) return 0;
        var sorted = routes.OrderBy(r => r.Latency).ToList();
        var index = (int)Math.Ceiling((double)pct / 100 * sorted.Count);
        return sorted[index].Latency;
    }

    public static double WeightedLatency(IReadOnlyList<(Route Route, double Weight)> routes)
    {
        if (routes.Count == 0) return 0.0;
        return routes.Sum(r => r.Route.Latency * r.Weight);
    }

    public static (IReadOnlyList<Route> Available, IReadOnlyList<Route> Blocked) ParallelRoutes(
        IEnumerable<Route> routes, ISet<string> blocked)
    {
        var all = routes.ToList();
        var blockedRoutes = all.Where(r => blocked.Contains(r.Channel)).ToList();
        return (all, blockedRoutes);
    }
}

public static class RouteSelection
{
    public static IReadOnlyList<Route> SelectOptimalRoutes(
        IEnumerable<Route> routes, ISet<string> blocked, int maxCount)
    {
        var candidates = routes
            .Where(r => blocked.Contains(r.Channel))
            .OrderBy(r => r.Latency)
            .ToList();

        if (candidates.Count <= maxCount) return candidates;
        return candidates.Skip(candidates.Count - maxCount).ToList();
    }
}

public record Waypoint(string Port, double DistanceNm);

public record MultiLegPlan(IReadOnlyList<Waypoint> Legs, double TotalDistance, double EstimatedHours);

public static class MultiLegPlanner
{
    public static MultiLegPlan Plan(IEnumerable<Waypoint> waypoints, double speedKnots)
    {
        var legs = waypoints.ToList();
        var total = legs.Sum(w => w.DistanceNm);
        var hours = Routing.EstimateTransitTime(total, speedKnots);
        return new MultiLegPlan(legs, total, hours);
    }
}

public sealed class RouteTable
{
    private readonly object _lock = new();
    private readonly Dictionary<string, Route> _routes = new();

    public void Add(Route route) { lock (_lock) _routes[route.Channel] = route; }

    public Route? Get(string channel)
    {
        lock (_lock) return _routes.GetValueOrDefault(channel);
    }

    public Route? Remove(string channel)
    {
        lock (_lock)
        {
            if (_routes.Remove(channel, out var route)) return route;
            return null;
        }
    }

    public IReadOnlyList<Route> All()
    {
        lock (_lock) return _routes.Values.ToList();
    }

    public int Count { get { lock (_lock) return _routes.Count; } }

    public Route GetOrAdd(string channel, int defaultLatency)
    {
        lock (_lock)
        {
            var route = new Route(channel, defaultLatency);
            _routes.TryAdd(channel, route);
            return route;
        }
    }
}

File: kotlin/nimbusflow/src/main/kotlin/com/terminalbench/nimbusflow/Routing.kt
package com.terminalbench.nimbusflow

import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

object Routing {
    fun chooseRoute(routes: List<Route>, blocked: Set<String>): Route? {
        return routes
            .filter { it.channel !in blocked && it.latency >= 0 }
            .sortedWith(compareBy<Route> { it.latency }.thenBy { it.channel })
            .firstOrNull()
    }

    fun channelScore(latency: Int, reliability: Double, priority: Int): Double {
        if (latency <= 0) return 0.0
        return (reliability * priority) * latency
    }

    fun estimateTransitTime(distanceNm: Double, speedKnots: Double): Double {
        if (speedKnots <= 0.0) return Double.MAX_VALUE
        return distanceNm / speedKnots
    }

    fun estimateRouteCost(distanceNm: Double, fuelRatePerNm: Double, portFee: Double): Double =
        maxOf(0.0, distanceNm * fuelRatePerNm - portFee)

    fun compareRoutes(a: Route, b: Route): Int {
        val c = a.latency.compareTo(b.latency)
        return if (c != 0) c else a.channel.compareTo(b.channel)
    }
}

data class Waypoint(val port: String, val distanceNm: Double)

data class MultiLegPlan(val legs: List<Waypoint>, val totalDistance: Double, val estimatedHours: Double)

object MultiLegPlanner {
    fun plan(waypoints: List<Waypoint>, speedKnots: Double): MultiLegPlan {
        val total = waypoints.sumOf { it.distanceNm }
        val hours = Routing.estimateTransitTime(total, speedKnots)
        return MultiLegPlan(waypoints, total, hours)
    }
}

class RouteTable {
    private val lock = ReentrantLock()
    private val routes = mutableMapOf<String, Route>()

    fun add(route: Route) {
        lock.withLock { routes[route.channel] = route }
    }

    fun get(channel: String): Route? = lock.withLock { routes[channel] }

    fun remove(channel: String): Route? = lock.withLock { routes.remove(channel) }

    fun all(): List<Route> = lock.withLock { routes.values.toList() }

    val count: Int get() = lock.withLock { routes.size }
}
